toolkit/smart-contracts/offchain/src/multisig.rs

```diff
@@ -0,0 +1,160 @@
+	pub private_key: CardanoPaymentSigningKey,
+	pub funded_by_tx: [u8; 32],
+}
+
+impl From<TemporaryWallet> for TemporaryWalletData {
+	fn from(value: TemporaryWallet) -> Self {
+		TemporaryWalletData {
+			address: value.address.to_bech32(None).unwrap(),
+			private_key: value.private_key.to_bytes().into(),
+			funded_by_tx: McTxHash(value.funded_by_tx),
+		}
+	}
+}
+
+fn serialize_as_conway_tx<S>(tx_bytes: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	let json = serde_json::json!({
+		"type": "Tx ConwayEra",
+		"description": "",
+		"cborHex": hex::encode(tx_bytes)
+	});
+	json.serialize(serializer)
+}
+
+impl MultiSigSmartContractResult {
+	pub fn tx_submitted(hash: [u8; 32]) -> Self {
+		Self::TransactionSubmitted(McTxHash(hash))
+	}
+}
+
+pub(crate) async fn create_temporary_wallet<
+	T: QueryLedgerState + Transactions + QueryNetwork + QueryUtxoByUtxoId,
+	A: AwaitTx,
+>(
+	ctx: &TransactionContext,
+	client: &T,
+	await_tx: &A,
+) -> anyhow::Result<TemporaryWallet> {
+	let private_key = CardanoPaymentSigningKey(PrivateKey::generate_ed25519()?);
+	log::info!(
+		"Temporary wallet private key: {}. Store this key for eventual re-claim of tokens",
+		private_key.0.to_hex()
+	);
+	let address = key_hash_address(&private_key.0.to_public().hash(), ctx.network);
+
+	let mut funding_tx_builder = TransactionBuilder::new(&get_builder_config(ctx)?);
+	// TODO: ETCM-9627 - estimate tokens needed for the transaction
+	funding_tx_builder
+		.add_output(&TransactionOutput::new(&address, &Value::new(&5_000_000u32.into())))?;
+	let funding_tx = funding_tx_builder.balance_update_and_build(ctx)?;
+	let funding_tx_result = client.submit_transaction(&ctx.sign(&funding_tx).to_bytes()).await?;
+	let funded_by_tx = funding_tx_result.transaction.id;
+	await_tx.await_tx_output(client, UtxoId::new(funded_by_tx, 0)).await?;
+	let address_str: String = address.to_bech32(None)?;
+	log::info!(
+		"Founded temporary wallet {} with 5 ADA in transaction: {}",
+		&address_str,
+		&hex::encode(funded_by_tx)
+	);
+	Ok(TemporaryWallet { address, private_key, funded_by_tx })
+}
+
+/// If the chain has real MultiSig governance it:
+/// * creates a temporary wallet
+/// * sends 5 ADA from the payment wallet (subject of change)
+/// * creates a transaction that would be paid from the temporary wallet, signed by both wallets.
+/// If the chain has single key governance it creates and submits transaction paid by and signed by the payment wallet.
+pub(crate) async fn multisig_process<F, T, A>(
+	governance_data: &GovernanceData,
+	payment_ctx: &TransactionContext,
+	make_tx: F,
+	client: &T,
+	await_tx: &A,
+) -> anyhow::Result<MultiSigSmartContractResult>
+where
+	F: Fn(Costs, &TransactionContext) -> anyhow::Result<Transaction>,
+	T: QueryLedgerState + Transactions + QueryNetwork + QueryUtxoByUtxoId,
+	A: AwaitTx,
+{
+	if governance_data.policy.is_single_key_policy_for(&payment_ctx.payment_key_hash()) {
```

Comment:
Maybe the function body could be refactored a bit to make the 2 cases more obvious and extract the parts that happen in both branches?

```
if governance_data.policy.is_single_key_policy_for(&payment_ctx.payment_key_hash()) {
   ...
   signed_tx = create_signed_transaction(..)
}
else {
  signed_tx_by_caller = create_signed_transaction(..)
   ...
}
```

Just a thought to maybe make the function a bit easier go grasp and making the two logic branches more obvious

Reply:
This is done.  Costed some lines of code, because the functions signatures are almost as long as function, but I think it is better after change.